Upgrading from 24.6.7 -> 24.7.0 now throws exception when trying to add elements to HorizontalLayout in certain async scenario

### Describe the bug

Code:
```
public void apply(final HorizontalLayout layout) {
        layout.add(this.components.toArray(Component[]::new)); // <-- exception here
        layout.setSpacing(spacing);
        layout.setPadding(padding);
        if (spacing) {
            layout.setSpacing(false);
            layout.getThemeList().set(spacingTheme, true);
        }
        layout.addClassNames(this.classNames.toArray(String[]::new));
        if (styleWidth != null) {
            layout.getStyle().setWidth(styleWidth);
        }
        if (sizeFull) {
            layout.setSizeFull();
        }
    }
```
I checked `this.components` and it is a non-null array with non-null components (icons, spans).

ConsoleNotificationService:
```
@Getter
@Service
public class ConsoleNotificationsService {

    private List<NewsBO> currentNotifications = Collections.emptyList();
    private final List<UI> listOfUi = new ArrayList<>();

    /**
     * Periodically checks for news updates of prioriy 3 CONSOLE - {@link NewsPriority}
     * Then, shows notifications for each registered UI if not already shown (this ensures showing notification without navigation)
     */
    @Scheduled(timeUnit = TimeUnit.SECONDS, initialDelay = 5, fixedRate = 15)
    public void updateConsoleNotifications() {
        val now = LocalDateTime.now();
        currentNotifications = getCurrentNotifications(now);
        listOfUi.removeIf(ui -> !ui.isAttached() || ui.isClosing());
        for (val ui : listOfUi) {
            val session = ui.getSession();
            ui.access(() -> {
                val shownNotif = session.getAttribute(Constants.SessionAttributes.SHOWN_CONSOLE_NOTIFICATIONS);
                assert shownNotif != null;
                synchronized (shownNotif) {
                    for (val news : currentNotifications) {
                        val shownNotifSet = (Set<Long>) shownNotif;
                        if (!shownNotifSet.contains(news.getId())) {
                            MainLayout.createConsoleNotification(news).open();
                            shownNotifSet.add(news.getId());
                        }
                    }
                }
                ui.push();
            });
        }
        ...
    } 
```
This service periodically checks for messages in db and show them in ui if there are any, and updates some session attributes.
List of uis is updated with ui init listener.

Stack trace:
```
java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: Cannot insert element with index 4 when there are 3 children
	at java.base/java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.base/java.util.concurrent.FutureTask.get(FutureTask.java:191)
	at com.vaadin.flow.server.FutureAccess.get(FutureAccess.java:62)
	at com.vaadin.flow.server.VaadinService.runPendingAccessTasks(VaadinService.java:2176)
	at com.vaadin.flow.component.UI.push(UI.java:751)
	at sk.bozpo.eboz.vaadin.config.ConsoleNotificationsService.lambda$updateConsoleNotifications$61ce4116$1(ConsoleNotificationsService.java:51)
	at com.vaadin.flow.component.UI.accessSynchronously(UI.java:498)
	at com.vaadin.flow.component.UI$2.execute(UI.java:573)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:317)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at --- Async.Stack.Trace --- (captured by IntelliJ IDEA debugger)
	at java.base/java.util.concurrent.FutureTask.<init>(FutureTask.java:151)
	at com.vaadin.flow.server.FutureAccess.<init>(FutureAccess.java:44)
	at com.vaadin.flow.server.VaadinService.accessSession(VaadinService.java:2101)
	at com.vaadin.flow.server.VaadinSession.access(VaadinSession.java:1059)
	at com.vaadin.flow.component.UI.access(UI.java:570)
	at com.vaadin.flow.component.UI.access(UI.java:553)
	at sk.bozpo.eboz.vaadin.config.ConsoleNotificationsService.updateConsoleNotifications(ConsoleNotificationsService.java:39)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.runInternal(ScheduledMethodRunnable.java:130)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.lambda$run$2(ScheduledMethodRunnable.java:124)
	at io.micrometer.observation.Observation.observe(Observation.java:498)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:124)
	at org.springframework.scheduling.config.Task$OutcomeTrackingRunnable.run(Task.java:85)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.runAndReset$$$capture(FutureTask.java:358)
	at java.base/java.util.concurrent.FutureTask.runAndReset(FutureTask.java)
	at --- Async.Stack.Trace --- (captured by IntelliJ IDEA debugger)
	at java.base/java.util.concurrent.FutureTask.<init>(FutureTask.java:151)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.<init>(ScheduledThreadPoolExecutor.java:227)
	at java.base/java.util.concurrent.ScheduledThreadPoolExecutor.scheduleAtFixedRate(ScheduledThreadPoolExecutor.java:629)
	at org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler.scheduleAtFixedRate(ThreadPoolTaskScheduler.java:439)
	at org.springframework.scheduling.config.TaskSchedulerRouter.scheduleAtFixedRate(TaskSchedulerRouter.java:121)
	at org.springframework.scheduling.config.ScheduledTaskRegistrar.scheduleFixedRateTask(ScheduledTaskRegistrar.java:551)
	at org.springframework.scheduling.config.ScheduledTaskRegistrar.scheduleTasks(ScheduledTaskRegistrar.java:446)
	at org.springframework.scheduling.config.ScheduledTaskRegistrar.afterPropertiesSet(ScheduledTaskRegistrar.java:421)
	at org.springframework.scheduling.annotation.ScheduledAnnotationBeanPostProcessor.finishRegistration(ScheduledAnnotationBeanPostProcessor.java:269)
	at org.springframework.scheduling.annotation.ScheduledAnnotationBeanPostProcessor.onApplicationEvent(ScheduledAnnotationBeanPostProcessor.java:668)
	at org.springframework.scheduling.annotation.ScheduledAnnotationBeanPostProcessor.onApplicationEvent(ScheduledAnnotationBeanPostProcessor.java:113)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.doInvokeListener(SimpleApplicationEventMulticaster.java:185)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.invokeListener(SimpleApplicationEventMulticaster.java:178)
	at org.springframework.context.event.SimpleApplicationEventMulticaster.multicastEvent(SimpleApplicationEventMulticaster.java:156)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:454)
	at org.springframework.context.support.AbstractApplicationContext.publishEvent(AbstractApplicationContext.java:387)
	at org.springframework.context.support.AbstractApplicationContext.finishRefresh(AbstractApplicationContext.java:1009)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:630)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:146)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:752)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:439)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:318)
	at org.springframework.boot.builder.SpringApplicationBuilder.run(SpringApplicationBuilder.java:149)
	at sk.bozpo.eboz.Application.main(Application.java:32)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:50)
Caused by: java.lang.IllegalArgumentException: Cannot insert element with index 4 when there are 3 children
	at com.vaadin.flow.dom.Node.insertChild(Node.java:358)
	at com.vaadin.flow.dom.Node.insertChild(Node.java:340)
	at com.vaadin.flow.component.orderedlayout.HorizontalLayout.lambda$addToStart$2(HorizontalLayout.java:382)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.accept(ForEachOps.java:184)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ForEachOps$ForEachOp.evaluateSequential(ForEachOps.java:151)
	at java.base/java.util.stream.ForEachOps$ForEachOp$OfRef.evaluateSequential(ForEachOps.java:174)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.forEach(ReferencePipeline.java:596)
	at com.vaadin.flow.component.orderedlayout.HorizontalLayout.addToStart(HorizontalLayout.java:381)
	at com.vaadin.flow.component.orderedlayout.HorizontalLayout.addToStart(HorizontalLayout.java:358)
	at com.vaadin.flow.component.orderedlayout.HorizontalLayout.add(HorizontalLayout.java:323)
	at sk.bozpo.eboz.moduly.core.CustomizableHorizontalLayout$Builder.apply(CustomizableHorizontalLayout.java:205)
	at sk.bozpo.eboz.moduly.core.CustomizableHorizontalLayout$Builder.build(CustomizableHorizontalLayout.java:200)
	at sk.bozpo.eboz.vaadin.config.VaadinErrorHandler.showErrorDialog(VaadinErrorHandler.java:49)
	at sk.bozpo.eboz.vaadin.config.VaadinErrorHandler.lambda$error$d3720659$1(VaadinErrorHandler.java:32)
	at com.vaadin.flow.component.UI.accessSynchronously(UI.java:498)
	at com.vaadin.flow.component.UI$2.execute(UI.java:573)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:572)
	at java.base/java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:317)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java)
	at --- Async.Stack.Trace --- (captured by IntelliJ IDEA debugger)
	at java.base/java.util.concurrent.FutureTask.<init>(FutureTask.java:151)
	at com.vaadin.flow.server.FutureAccess.<init>(FutureAccess.java:44)
	at com.vaadin.flow.server.VaadinService.accessSession(VaadinService.java:2101)
	at com.vaadin.flow.server.VaadinSession.access(VaadinSession.java:1059)
	at com.vaadin.flow.component.UI.access(UI.java:570)
	at com.vaadin.flow.component.UI.access(UI.java:553)
	at sk.bozpo.eboz.vaadin.config.VaadinErrorHandler.error(VaadinErrorHandler.java:31)
	at com.vaadin.flow.server.communication.ServerRpcHandler.callErrorHandler(ServerRpcHandler.java:587)
	at com.vaadin.flow.server.communication.ServerRpcHandler.handleInvocationData(ServerRpcHandler.java:573)
	at com.vaadin.flow.server.communication.ServerRpcHandler.lambda$handleInvocations$6(ServerRpcHandler.java:549)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1596)
	at com.vaadin.flow.server.communication.ServerRpcHandler.handleInvocations(ServerRpcHandler.java:549)
	at com.vaadin.flow.server.communication.ServerRpcHandler.handleRpc(ServerRpcHandler.java:376)
	at com.vaadin.flow.server.communication.UidlRequestHandler.synchronizedHandleRequest(UidlRequestHandler.java:138)
	at com.vaadin.flow.server.SynchronizedRequestHandler.handleRequest(SynchronizedRequestHandler.java:63)
	at com.vaadin.flow.server.VaadinService.handleRequest(VaadinService.java:1664)
	at com.vaadin.flow.server.VaadinServlet.service(VaadinServlet.java:398)
	at com.vaadin.flow.spring.SpringServlet.service(SpringServlet.java:106)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:195)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.apache.catalina.core.ApplicationDispatcher.invoke(ApplicationDispatcher.java:633)
	at org.apache.catalina.core.ApplicationDispatcher.processRequest(ApplicationDispatcher.java:409)
	at org.apache.catalina.core.ApplicationDispatcher.doForward(ApplicationDispatcher.java:304)
	at org.apache.catalina.core.ApplicationDispatcher.forward(ApplicationDispatcher.java:268)
	at org.springframework.web.servlet.mvc.ServletForwardingController.handleRequestInternal(ServletForwardingController.java:142)
	at org.springframework.web.servlet.mvc.AbstractController.handleRequest(AbstractController.java:178)
	at org.springframework.web.servlet.mvc.SimpleControllerHandlerAdapter.handle(SimpleControllerHandlerAdapter.java:51)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1089)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:979)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1014)
	at org.springframework.web.servlet.FrameworkServlet.doPost(FrameworkServlet.java:914)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:590)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:885)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:658)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:195)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:51)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.RequestContextFilter.doFilterInternal(RequestContextFilter.java:100)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.FormContentFilter.doFilterInternal(FormContentFilter.java:93)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:201)
	at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:116)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:140)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:167)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:90)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:483)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:115)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:93)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:74)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:344)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:397)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:63)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:905)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1743)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1190)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:63)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

App configuration:
```
@SpringBootApplication
@Theme(value = "myapp")
@PWA(name = "Eboz", shortName = "Eboz", iconPath = "icons/favicon.png", offlinePath = "offline.html")
@Push(value = PushMode.MANUAL, transport = Transport.LONG_POLLING)
@EnableAsync
@EnableScheduling
public class Application extends SpringBootServletInitializer implements AppShellConfigurator {

    // dev only
    public static void main(final String[] args) throws Exception {
        new SpringApplicationBuilder(Application.class)
                .listeners(new DatabasePropertiesSetup())
                .run(args);
    }

    // when deployed to servlet container
    @Override
    protected SpringApplicationBuilder configure(final SpringApplicationBuilder application) {
        return application.listeners(new DatabasePropertiesSetup())
                .sources(Application.class);
    }
}
```

### Expected-behavior

Will not throw exception (as in 24.6.7) 

### Reproduction

Reproduction might be quite complex. I will update more code if needed.

### System Info

Win 10
This error occurs when changed vaadin version from 24.6.7 to 24.7.0
Firefox
IntelliJ IDEA 2024.3.5 (Community Edition)

--- a/vaadin-ordered-layout-flow-parent/vaadin-ordered-layout-flow/src/main/java/com/vaadin/flow/component/orderedlayout/HorizontalLayout.java
+++ b/vaadin-ordered-layout-flow-parent/vaadin-ordered-layout-flow/src/main/java/com/vaadin/flow/component/orderedlayout/HorizontalLayout.java
@@ -379,7 +379,11 @@ public class HorizontalLayout extends Component implements ThemableLayout,
                 .map(component -> Objects.requireNonNull(component,
                         "Component to add cannot be null"))
                 .forEach((component) -> {
-                    getElement().insertChild(itemCounter.getAndIncrement(),
+                    var isChild = getElement()
+                            .equals(component.getElement().getParent());
+                    getElement().insertChild(
+                            isChild ? itemCounter.get()
+                                    : itemCounter.getAndIncrement(),
                             component.getElement());
                 });
     }
@@ -429,7 +433,12 @@ public class HorizontalLayout extends Component implements ThemableLayout,
                     }
 
                     component.getElement().setAttribute("slot", "middle");
-                    getElement().insertChild(itemCounter.getAndIncrement(),
+
+                    var isChild = getElement()
+                            .equals(component.getElement().getParent());
+                    getElement().insertChild(
+                            isChild ? itemCounter.get()
+                                    : itemCounter.getAndIncrement(),
                             component.getElement());
                 });
 

--- a/vaadin-ordered-layout-flow-parent/vaadin-ordered-layout-flow/src/test/java/com/vaadin/flow/component/orderedlayout/tests/HorizontalLayoutSlotsTest.java
+++ b/vaadin-ordered-layout-flow-parent/vaadin-ordered-layout-flow/src/test/java/com/vaadin/flow/component/orderedlayout/tests/HorizontalLayoutSlotsTest.java
@@ -305,6 +305,54 @@ public class HorizontalLayoutSlotsTest {
     }
 
     @Test
+    public void add_sameComponentAddedTwice_doesNotThrow() {
+        Div div1 = new Div();
+        Div div2 = new Div();
+
+        layout.add(div1);
+        layout.add(div1, div2);
+
+        Assert.assertEquals(div1, layout.getComponentAt(0));
+        Assert.assertEquals(div2, layout.getComponentAt(1));
+    }
+
+    @Test
+    public void add_sameComponentAddedTwice_changeOrder_doesNotThrow() {
+        Div div1 = new Div();
+        Div div2 = new Div();
+
+        layout.add(div1);
+        layout.add(div2, div1);
+
+        Assert.assertEquals(div2, layout.getComponentAt(0));
+        Assert.assertEquals(div1, layout.getComponentAt(1));
+    }
+
+    @Test
+    public void addToMiddle_sameComponentAddedTwice_doesNotThrow() {
+        Div div1 = new Div();
+        Div div2 = new Div();
+
+        layout.addToMiddle(div1);
+        layout.addToMiddle(div1, div2);
+
+        Assert.assertEquals(div1, layout.getComponentAt(0));
+        Assert.assertEquals(div2, layout.getComponentAt(1));
+    }
+
+    @Test
+    public void addToMiddle_sameComponentAddedTwice_changeOrder_doesNotThrow() {
+        Div div1 = new Div();
+        Div div2 = new Div();
+
+        layout.addToMiddle(div1);
+        layout.addToMiddle(div2, div1);
+
+        Assert.assertEquals(div2, layout.getComponentAt(0));
+        Assert.assertEquals(div1, layout.getComponentAt(1));
+    }
+
+    @Test
     public void addComponentAtIndex_firstComponentAdded() {
         Div div1 = new Div();
         layout.addComponentAtIndex(0, div1);

